From 6968ebf1006c3fbcb24379f16e085544106849ae Mon Sep 17 00:00:00 2001
From: Stefan Eissing <stefan@eissing.org>
Date: Wed, 28 Oct 2026 12:58:37 +0100
Subject: [PATCH] connection close: additional checks

Connection pool: *never* free a connection that has still transfers
attached, abort flag set or not.
Curl_connect: on error, close the selected connection only if
it was not reused and has no other transfers attached.

Fixes #23629
Closes #23646
---
 lib/conncache.c | 7 +++----
 lib/url.c       | 8 +++++---
 2 files changed, 8 insertions(+), 7 deletions(-)

diff --git a/lib/conncache.c b/lib/conncache.c
index 0e41a1a..105f847 100644
--- a/lib/conncache.c
+++ b/lib/conncache.c
@@ -402,10 +402,9 @@ static void cpool_conn_close(struct cpool *cpool,
   if(!cpool)
     return;
 
-  /* If this connection is not marked to force-close, leave it open if there
-   * are other users of it */
-  if(CONN_INUSE(conn) && !aborted) {
-    DEBUGASSERT(0); /* does this ever happen? */
+  /* If this connection is still used by transfers, we cannot close it. */
+  if(CONN_INUSE(conn)) {
+    DEBUGASSERT(0); /* we'd like to know when this happens */
     DEBUGF(infof(data, "conn terminate when inuse: %u", conn->attached_xfers));
     return;
   }
diff --git a/lib/url.c b/lib/url.c
index 7c315ec..a137fa7 100644
--- a/lib/url.c
+++ b/lib/url.c
@@ -2527,10 +2527,12 @@ out:
     DEBUGASSERT(!conn);
 
   if(result && conn) {
-    /* We are not allowed to return failure with memory left allocated in the
-       connectdata struct, free those here */
+    /* Detach on error and close the connection if
+     * - it was not reused (meaning, ok before we picked it)
+     * - there are no other transfers attached to it */
     Curl_detach_connection(data);
-    Curl_conn_close(data, conn, TRUE);
+    if(!conn->bits.reuse && !CONN_INUSE(conn))
+      Curl_conn_close(data, conn, TRUE);
   }
 
   return result;
-- 
2.53.0

